From db5953a096569a9679b7937f3ecca6fb2c0667d9 Mon Sep 17 00:00:00 2001
From: Marcin Wojtas <mw@semihalf.com>
Date: Fri, 15 Feb 2019 23:00:45 +0100
Subject: [PATCH 1002/1051] net: mvneta: declare array for skbs in RX SWBM as
 variable length

It turned out that creating an array for collecting received
skbs with a macro NAPI_POLL_WEIGHT instead of a budget
value passed to mvneta_poll result in a performance
drop due to worse Icache utilisation.

Measurements with command during 64B L3FWD test:
perf record -e L1-icache-load-misses -C 0 sleep 10

* struct sk_buff *rcvd_skbs[budget];
  - ~91M Icache misses

* struct sk_buff *rcvd_skbs[NAPI_POLL_WEIGHT];
  - ~123M Icache misses

Disasembling shows a different code structure with the constant
size of the array, which prevents improving Icache utilization.
Restore original version of the improvement, using variable length
of the skb array.

Change-Id: I4afc3d5f234fa414a0770056febc6bbeac9ab682
Signed-off-by: Marcin Wojtas <mw@semihalf.com>
Reviewed-on: https://sj1git1.cavium.com/4179
Tested-by: sa_ip-sw-jenkins
Reviewed-by: Stefan Chulski <Stefan.Chulski@cavium.com>
[Kevin: The original patch got from Marvell sdk10.0_19.06]
Signed-off-by: Kevin Hao <kexin.hao@windriver.com>
---
 drivers/net/ethernet/marvell/mvneta.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/drivers/net/ethernet/marvell/mvneta.c b/drivers/net/ethernet/marvell/mvneta.c
index 71488c3a03c8..6141c26d4506 100644
--- a/drivers/net/ethernet/marvell/mvneta.c
+++ b/drivers/net/ethernet/marvell/mvneta.c
@@ -1946,7 +1946,7 @@ static int mvneta_rx_swbm(struct napi_struct *napi,
 			  struct mvneta_port *pp, int budget,
 			  struct mvneta_rx_queue *rxq)
 {
-	struct sk_buff *rcvd_skbs[NAPI_POLL_WEIGHT];
+	struct sk_buff *rcvd_skbs[budget];
 	struct net_device *dev = pp->dev;
 	int rx_todo, rx_proc;
 	int refill = 0, i = 0;
-- 
2.17.1

